src/libraries/System.Private.CoreLib/src/System/Globalization/Ordinal.Utf8.cs

```diff
@@ -0,0 +1,261 @@
+
+                // A 32-bit test - even with the bit-twiddling here - is more efficient than a 64-bit test.
+                ulong temp = valueAu64 | valueBu64;
+                if (!Utf8Utility.AllBytesInUInt32AreAscii((uint)temp | (uint)(temp >> 32)))
+                {
+                    goto NonAscii64; // one of the inputs contains non-ASCII data
+                }
+
+                // Generally, the caller has likely performed a first-pass check that the input strings
+                // are likely equal. Consider a dictionary which computes the hash code of its key before
+                // performing a proper deep equality check of the string contents. We want to optimize for
+                // the case where the equality check is likely to succeed, which means that we want to avoid
+                // branching within this loop unless we're about to exit the loop, either due to failure or
+                // due to us running out of input data.
+
+                if (!Utf8Utility.UInt64OrdinalIgnoreCaseAscii(valueAu64, valueBu64))
+                {
+                    return false;
+                }
+
+                byteOffset += 8;
+                length -= 8;
+            }
+#endif
+            uint valueAu32 = 0;
+            uint valueBu32 = 0;
+            // Read 4 chars (32 bits) at a time from each string
+#if TARGET_64BIT
+            if ((uint)length >= 4)
+#else
+            while ((uint)length >= 4)
+#endif
+            {
+                valueAu32 = Unsafe.ReadUnaligned<uint>(ref Unsafe.AddByteOffset(ref charA, byteOffset));
+                valueBu32 = Unsafe.ReadUnaligned<uint>(ref Unsafe.AddByteOffset(ref charB, byteOffset));
+
+                if (!Utf8Utility.AllBytesInUInt32AreAscii(valueAu32 | valueBu32))
+                {
+                    goto NonAscii32; // one of the inputs contains non-ASCII data
+                }
+
+                // Generally, the caller has likely performed a first-pass check that the input strings
+                // are likely equal. Consider a dictionary which computes the hash code of its key before
+                // performing a proper deep equality check of the string contents. We want to optimize for
+                // the case where the equality check is likely to succeed, which means that we want to avoid
+                // branching within this loop unless we're about to exit the loop, either due to failure or
+                // due to us running out of input data.
+
+                if (!Utf8Utility.UInt32OrdinalIgnoreCaseAscii(valueAu32, valueBu32))
+                {
+                    return false;
+                }
+
+                byteOffset += 4;
+                length -= 4;
+            }
+
+            if (length != 0)
+            {
+                // We have 1, 2, or 3 bytes remaining. We want to backtrack
+                // so we read exactly 4 bytes and then do one final iteration.
```

Comment:
Why is it safe to read 4 bytes if there's only at most 3 remaining?

Reply:
Ah, right. I was thinking we're in a path that has already checked and asserted we have at least 4 characters, so backtracking would be safe.

That is not actually the case so this optimization can't be done without an additional check.